packages/datetime/src/dateRangeInput.tsx

```diff
@@ -370,28 +410,45 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
     }
 
     private getSelectedRange = () => {
-        return [this.state.selectedStart, this.state.selectedEnd].map((selectedBound?: moment.Moment) => {
+        const momentDateRange = [this.state.selectedStart, this.state.selectedEnd] as MomentDateRange;
+        const [startDate, endDate] = momentDateRange.map((selectedBound?: moment.Moment) => {
             return this.isMomentValidAndInRange(selectedBound)
                 ? fromMomentToDate(selectedBound)
                 : undefined;
-        }) as DateRange;
+        });
+        // show only the start date if the dates overlap
+        // TODO: add different handling for the === case once
+        // allowSingleDayRange is implemented
+        return [startDate, (startDate >= endDate) ? null : endDate] as DateRange;
     }
 
-    private getStartInputDisplayString = () => {
-        return this.state.isStartInputFocused
-            ? this.state.startInputString
-            : this.getFormattedDateString(this.state.selectedStart);
-    }
+    private getInputDisplayString = (boundary: DateRangeBoundary) => {
+        const { values } = this.getStateKeysAndValuesForBoundary(boundary);
+        const { isInputFocused, inputString, selectedValue } = values;
 
-    private getEndInputDisplayString = () => {
-        return this.state.isEndInputFocused
-            ? this.state.endInputString
-            : this.getFormattedDateString(this.state.selectedEnd);
+        if (isInputFocused) {
+            return inputString;
+        } else {
+            if (isMomentNull(selectedValue)) {
+                return "";
+            } else if (!this.isMomentInRange(selectedValue)) {
+                return this.props.outOfRangeMessage;
+            } else if (this.doBoundariesOverlap(boundary, selectedValue) && boundary === DateRangeBoundary.END) {
+                return this.props.overlappingDatesMessage;
+            } else {
+                return this.getFormattedDateString(selectedValue);
+            }
+        }
     }
 
     private getFormattedDateString = (momentDate: moment.Moment) => {
         if (isMomentNull(momentDate)) {
             return "";
+        } else if (!momentDate.isValid()) {
+            // moment.js formats invalid dates as "Invalid date", which may or
+            // may not match the invalidDateMessage. let's just always show the
+            // custom invalidDateMessage for invalid dates, for consistency.
```

Comment:
not sure this comment is necessary. i wouldn't feel slighted or confused if it were removed. providing the prop seems strictly better than relying on moment.